askama_derive/src/input.rs

```diff
@@ -25,106 +37,36 @@ impl<'a> TemplateInput<'a> {
-                        panic!("must specify 'source' or 'path', not both");
-                    }
-                    source = Some(Source::Path(s.value()));
-                } else {
-                    panic!("template path must be string literal");
-                }
-            } else if pair.path.is_ident("source") {
-                if let syn::Lit::Str(ref s) = pair.lit {
-                    if source.is_some() {
-                        panic!("must specify 'source' or 'path', not both");
-                    }
-                    source = Some(Source::Source(s.value()));
-                } else {
-                    panic!("template source must be string literal");
-                }
-            } else if pair.path.is_ident("print") {
-                if let syn::Lit::Str(ref s) = pair.lit {
-                    print = s.value().into();
-                } else {
-                    panic!("print value must be string literal");
-                }
-            } else if pair.path.is_ident("escape") {
-                if let syn::Lit::Str(ref s) = pair.lit {
-                    escaping = Some(s.value());
-                } else {
-                    panic!("escape value must be string literal");
-                }
-            } else if pair.path.is_ident("ext") {
-                if let syn::Lit::Str(ref s) = pair.lit {
-                    ext = Some(s.value());
-                } else {
-                    panic!("ext value must be string literal");
-                }
-            } else if pair.path.is_ident("syntax") {
-                if let syn::Lit::Str(ref s) = pair.lit {
-                    syntax = Some(s.value())
-                } else {
-                    panic!("syntax value must be string literal");
-                }
-            } else {
-                panic!(
-                    "unsupported attribute key '{}' found",
-                    pair.path.to_token_stream()
-                )
-            }
-        }
+        assert_ne!(
+            source.is_some(),
+            path.is_some(),
+            "One of path or source must exist and they are mutually exclusive",
+        );
+        assert_eq!(
+            source.is_some(),
+            ext.is_some(),
+            "must include 'ext' attribute when using 'source' attribute"
+        );
 
-        // Validate the `source` and `ext` value together, since they are
-        // related. In case `source` was used instead of `path`, the value
-        // of `ext` is merged into a synthetic `path` value here.
-        let source = source.expect("template path or source not found in attributes");
+        // Since 'source' and 'path' are related. In case `source` was used instead
+        // of `path`, the value of `ext` is merged into a synthetic `path` value here.
+        let source = source
+            .map(Source::Source)
+            .unwrap_or_else(|| Source::Path(path.unwrap()));
         let path = match (&source, &ext) {
```

Comment:
If you're going to do this, let's also move this login into a `Source::path()` method.